cmake_modules/FindGMP.cmake

```diff
@@ -3,17 +3,31 @@
 # GMP_INCLUDE_DIR - the GMP include directory
 # GMP_LIBRARIES - Libraries needed to use GMP
 
-if (GMP_INCLUDE_DIRS AND GMP_LIBRARIES)
-		# Already in cache, be silent
-		set(GMP_FIND_QUIETLY TRUE)
-endif (GMP_INCLUDE_DIRS AND GMP_LIBRARIES)
-
-find_path(GMP_INCLUDE_DIRS NAMES gmp.h )
-find_library(GMP_LIBRARIES NAMES gmp libgmp )
-find_library(GMPXX_LIBRARIES NAMES gmpxx libgmpxx )
-MESSAGE(STATUS "GMP libs: " ${GMP_LIBRARIES} " " ${GMPXX_LIBRARIES} )
+find_path(GMP_INCLUDE_DIR NAMES gmp.h )
+find_library(GMP_LIBRARY NAMES gmp libgmp )
+find_library(GMPXX_LIBRARY NAMES gmpxx libgmpxx )
+mark_as_advanced(GMP_INCLUDE_DIR GMP_LIBRARY GMPXX_LIBRARY)
+MESSAGE(STATUS "GMP libs: " ${GMP_LIBRARY} " " ${GMPXX_LIBRARY} )
 
 include(FindPackageHandleStandardArgs)
-FIND_PACKAGE_HANDLE_STANDARD_ARGS(GMP DEFAULT_MSG GMP_INCLUDE_DIRS GMP_LIBRARIES)
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(GMP DEFAULT_MSG GMP_INCLUDE_DIR GMP_LIBRARY GMPXX_LIBRARY)
+
+if(GMP_FOUND AND NOT TARGET GMP::GMP)
+    add_library(GMP::GMP_C UNKNOWN IMPORTED)
+    set_target_properties(GMP::GMP_C PROPERTIES
+        IMPORTED_LOCATION "${GMP_LIBRARY}"
+        INTERFACE_INCLUDE_DIRECTORIES "${GMP_INCLUDE_DIR}"
+    )
+    add_library(GMP::GMP_CXX UNKNOWN IMPORTED)
+    set_target_properties(GMP::GMP_CXX PROPERTIES
+        IMPORTED_LOCATION "${GMPXX_LIBRARY}"
+        #INTERFACE_INCLUDE_DIRECTORIES "${GMP_INCLUDE_DIR}"
+    )
 
-mark_as_advanced(GMP_INCLUDE_DIRS GMP_LIBRARIES GMPXX_LIBRARIES)
+    add_library(GMP::GMP INTERFACE IMPORTED)
+    set_property(TARGET GMP::GMP APPEND PROPERTY
+        INTERFACE_LINK_LIBRARIES GMP::GMP_C GMP::GMP_CXX
+        )
+    # the above command is for compatibillity with old versions of CMAKE, the version below is simpler, but works only from CMake 3.11 
```

Comment:
Referencing the above comment, add `include_directories(${GMP_INCLUDE_DIR})` somewhere here?

Reply:
I think I have an idea why the compilation might break in your case, although it is interesting that I was able to build OpenSMT in the cluster, with non-standard GMP location with no problems.
I think we should re-design how the main library is created. When I started I used this approach that each folder is compiled separately to a bunch of object files collected in an object library and later all these object files are linked together. Now I don't like this approach very much.
I am not sure if we can identify separate modules (targets) in OpenSMT that (at least in theory) would form standalone units without cyclic dependencies. Or we could keep just one main target and collect only source files from directories instead of object files. But I think I ran into a problem of how to create both static and shared library in that case.

Maybe we can wait with this PR a little bit, I will try to figure out something.

Reply:
I believe I have solved the problem to my liking. You can try compiling it again. Unfortunately, the CI has not run for the new commits, maybe after I rebase. Hopefully I have not missed anything.

Reply:
So far everything looks good to me.